File: samples/csharp_dotnetcore/54.SlackAdapterBot/Bots/EchoBot.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Adapters.Slack;
using Microsoft.Bot.Schema;
using Newtonsoft.Json;
using SlackAPI;
using Attachment = Microsoft.Bot.Schema.Attachment;

namespace SlackAdapterBot.Bots
{
    /// <summary>
    /// An EchoBot class that extends from the ActivityHandler.
    /// </summary>
    public class EchoBot : ActivityHandler
    {
        /// <summary>
        /// OnMessageActivityAsync method that returns an async Task.
        /// </summary>
        /// <param name="turnContext">turnContext of ITurnContext{T}, where T is an IActivity.</param>
        /// <param name="cancellationToken">cancellationToken propagates notifications that operations should be canceled.</param>
        /// <returns>A <see cref="Task{TResult}"/> representing the result of the asynchronous operation.</returns>
        protected override async Task OnMessageActivityAsync(ITurnContext<IMessageActivity> turnContext, CancellationToken cancellationToken)
        {
            await turnContext.SendActivityAsync(MessageFactory.Text($"Echo: {turnContext.Activity.Text}"), cancellationToken);
        }

        /// <summary>
        /// OnMessageActivityAsync method that returns an async Task.
        /// </summary>
        /// <param name="turnContext">turnContext of ITurnContext{T}, where T is an IActivity.</param>
        /// <param name="cancellationToken">cancellationToken propagates notifications that operations should be canceled.</param>
        /// <returns>A <see cref="Task{TResult}"/> representing the result of the asynchronous operation.</returns>
        protected override async Task OnEventActivityAsync(ITurnContext<IEventActivity> turnContext, CancellationToken cancellationToken)
        {
            if (turnContext.Activity.TryGetChannelData(out SlackRequestBody _))
            {
                if (turnContext.Activity.GetChannelData<SlackRequestBody>().Command == "/test")
                {
                    var interactiveMessage = MessageFactory.Attachment(
                        CreateInteractiveMessage(
                            Directory.GetCurrentDirectory() + @"\Resources\InteractiveMessage.json"));
                    await turnContext.SendActivityAsync(interactiveMessage, cancellationToken);
                }
            }

            if (turnContext.Activity.TryGetChannelData(out SlackEvent _))
            {
                if (turnContext.Activity.GetChannelData<SlackEvent>().SubType == "file_share")
                {
                    await turnContext.SendActivityAsync(MessageFactory.Text("Echo: I received an attachment"), cancellationToken);
                }
                else if (turnContext.Activity.GetChannelData<SlackEvent>().Message?.Attachments != null)
                {
                    await turnContext.SendActivityAsync(MessageFactory.Text("Echo: I received a link share"), cancellationToken);
                }
            }
        }

        /// <summary>
        /// OnMembersAddedAsync method that returns an async Task.
        /// </summary>
        /// <param name="membersAdded">membersAdded of IList{T}, where T is ChannelAccount.</param>
        /// <param name="turnContext">turnContext of ITurnContext{T}, where T is an IConversationUpdateActivity.</param>
        /// <param name="cancellationToken">cancellationToken propagates notifications that operations should be canceled.</param>
        /// <returns>A <see cref="Task{TResult}"/> representing the result of the asynchronous operation.</returns>
        protected override async Task OnMembersAddedAsync(IList<ChannelAccount> membersAdded, ITurnContext<IConversationUpdateActivity> turnContext, CancellationToken cancellationToken)
        {
            foreach (var member in membersAdded)
            {
                if (member.Id != turnContext.Activity.Recipient.Id)
                {
                    await turnContext.SendActivityAsync(MessageFactory.Text($"Hello and Welcome!"), cancellationToken);
                }
            }
        }

        private static Attachment CreateInteractiveMessage(string filePath)
        {
            var interactiveMessageJson = System.IO.File.ReadAllText(filePath);
            var adaptiveCardAttachment = JsonConvert.DeserializeObject<Block[]>(interactiveMessageJson);

            var blockList = adaptiveCardAttachment.ToList();

            var attachment = new Attachment
            {
                Content = blockList,
                ContentType = "application/json",
                Name = "blocks",
            };

            return attachment;
        }
    }
}

File: samples/python/60.slack-adapter/bots/echo_bot.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

import json
import os

from botbuilder.adapters.slack import SlackRequestBody, SlackEvent
from botbuilder.core import ActivityHandler, MessageFactory, TurnContext
from botbuilder.schema import ChannelAccount, Attachment


class EchoBot(ActivityHandler):
    async def on_members_added_activity(
        self, members_added: [ChannelAccount], turn_context: TurnContext
    ):
        for member in members_added:
            if member.id != turn_context.activity.recipient.id:
                await turn_context.send_activity("Hello and welcome!")

    async def on_message_activity(self, turn_context: TurnContext):
        return await turn_context.send_activity(
            MessageFactory.text(f"Echo: {turn_context.activity.text}")
        )

    async def on_event_activity(self, turn_context: TurnContext):
        body = turn_context.activity.channel_data
        if not body:
            return

        if isinstance(body, SlackRequestBody) and body.command == "/test":
            interactive_message = MessageFactory.attachment(
                self.__create_interactive_message(
                    os.path.join(os.getcwd(), "./resources/InteractiveMessage.json")
                )
            )
            await turn_context.send_activity(interactive_message)

        if isinstance(body, SlackEvent):
            if body.subtype == "file_share":
                await turn_context.send_activity("Echo: I received and attachment")
            elif body.message and body.message.attachments:
                await turn_context.send_activity("Echo: I received a link share")

    def __create_interactive_message(self, file_path: str) -> Attachment:
        # interactive_message_json = open(file_path).read()
        # adaptive_card_attachment = json.loads(interactive_message_json, cls=Block)
        with open(file_path, "rb") as in_file:
            adaptive_card_attachment = json.load(in_file)

        return Attachment(
            content=adaptive_card_attachment,
            content_type="application/json",
            name="blocks",
        )
